bfsscp/bfsscp.go

```diff
@@ -0,0 +1,440 @@
+// Package bfsscp abstracts an SSH/SCP workflow
+//
+// When imported, it registers both `scp://` and `ssh://` global scheme resolver and can be used like:
+//
+//   import (
+//     "github.com/bsm/bfs"
+//
+//     _ "github.com/bsm/bfs/bfsscp"
+//   )
+//
+//   func main() {
+//     ctx := context.Background()
+//     b, _ := bfs.Connect(ctx, "ssh://user:pass@hostname:22/path/to/root?tmpdir=%2Fcustom%2Ftmp")
+//     f, _ := b.Open(ctx, "file/within/root.txt")
+//     ...
+//   }
+//
+// bfs.Connect supports the following query parameters:
+//
+//   tmpdir - custom temp dir
+//
+package bfsscp
+
+import (
+	"context"
+	"io"
+	"io/ioutil"
+	"net"
+	"net/url"
+	"os"
+	"path"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/bmatcuk/doublestar"
+	"github.com/bsm/bfs"
+	"github.com/bsm/bfs/internal"
+	"github.com/kr/fs"
+	"github.com/pkg/sftp"
+	"go.uber.org/multierr"
+	"golang.org/x/crypto/ssh"
+)
+
+// register allows for registering more schemes with SCP
+func register(_ context.Context, u *url.URL) (bfs.Bucket, error) {
+	query := u.Query()
+	address := net.JoinHostPort(u.Host, u.Port())
+
+	username, password := "", ""
+	if u.User != nil {
+		username = u.User.Username()
+		password, _ = u.User.Password()
+	}
+
+	return New(address, &Config{
+		Username: username,
+		Password: password,
+		Prefix:   u.Path,
+		TempDir:  query.Get("tmpdir"),
+	})
+}
+
+func init() {
```

Comment:
Nit/cosmetics: I'd keep `init` at the top to see important bits right away (matter of taste though).